wasmtime/_memory.py

```diff
@@ -62,6 +64,64 @@ def data_ptr(self, store: Storelike) -> "ctypes._Pointer[c_ubyte]":
         """
         return ffi.wasmtime_memory_data(store._context, byref(self._memory))
 
+    def read(
+            self,
+            store: Storelike,
+            start: typing.Optional[int] = 0,
+            stop: typing.Optional[int] = None) -> bytearray:
+        """
+        Reads this memory starting from `start` and up to `stop`
+        and returns a copy of the contents as a byte array.
+
+        The indexing behavior of this method is similar to `list[start:stop]`
+        where negative starts can be used to read from the end, for example.
+        """
+        data_ptr = self.data_ptr(store)
+        size = self.data_len(store)
+        key = slice(start, stop, None)
+        start, stop, _ = key.indices(size)
+        val_size = stop - start
+        if val_size <= 0:
+            # return bytearray of size zero
+            return bytearray(0)
+        ptr_type = ctypes.c_ubyte * val_size
+        src_ptr = (ptr_type).from_address(ctypes.addressof(data_ptr.contents) + start)
+        return bytearray(src_ptr)
+
+    def write(
+            self,
+            store: Storelike,
+            value: typing.Any,
+            start: typing.Optional[int] = None,
+            stop: typing.Optional[int] = None) -> int:
+        """
+        write into a possibly large slice of memory
+        negative start, stop is allowed in a way similat to list slice mylist[-10:]
+        you can also ommit start in a way similar to mylist[:-10]
+        return number of bytes written
+        """
+        data_ptr = self.data_ptr(store)
+        size = self.data_len(store)
+        key = slice(start, stop)
+        start, stop, _ = key.indices(size)
+        # value must be bytearray ex. cast bytes() to bytearray
+        if not isinstance(value, array.array) and not isinstance(value, bytearray):
+            # value = array.array('B', value)
+            value = bytearray(value)
```

Comment:
Since this is a relatively low-level interface could this perhaps take `bytearray` directly? I think that would help with mypy typing as well since the argument would have a specific type.

Reply:
@muayyad-alsadi wrote:
> I want this function to be used with b"123" and with array.array() without casting
because it's very common to pass bytes to bytearrays and vice versa
...
and bytearray and array.array are done without copying
the pointer as seen by ctypes.memmove is equivalent in both bases

I agree with that point. But creating a `bytearray` from `bytes` will involve a copy, right? Because `bytearray` is mutable. So maybe just hint at it in the documentation that supplying `bytes` will copy twice?

Reply:
@j-blue-arz I've just added such a comment.